### PR TITLE
FIMS-24 Dismiss redis

### PR DESCRIPTION
**Pull request by datalek**

### CODE REVIEW

File: src/core/repositories/GrantRepository.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Option } from "fp-ts/Option";
+import * as TE from "fp-ts/TaskEither";
```
Comment (kin0992) [inline src/core/repositories/GrantRepository.ts:2]:
```suggestion
import { TaskEither } from "fp-ts/TaskEither";
```

File: src/core/repositories/GrantRepository.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ */
+export interface GrantRepository {
+  /**
+   * Given a grant create or update it.
+   */
+  readonly upsert: (grant: Grant) => TE.TaskEither<DomainError, Grant>;
+  /**
+   * Return the grant identfied with the given id.
+   */
+  readonly find: (id: GrantId) => TE.TaskEither<DomainError, Option<Grant>>;
```
Comment (kin0992) [inline src/core/repositories/GrantRepository.ts:16]:
```suggestion
  readonly upsert: (grant: Grant) => TaskEither<DomainError, Grant>;
  /**
   * Return the grant identified with the given id.
   */
  readonly find: (id: GrantId) => TaskEither<DomainError, Option<Grant>>;
```

File: src/core/repositories/InteractionRequestRepository.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { Option } from "fp-ts/Option";
+import * as TE from "fp-ts/TaskEither";
```
Comment (kin0992) [inline src/core/repositories/InteractionRequestRepository.ts:2]:
```suggestion
import { TaskEither } from "fp-ts/TaskEither";
```

File: src/core/repositories/InteractionRequestRepository.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+   */
+  readonly remove: (
+    id: InteractionRequestId
+  ) => TE.TaskEither<DomainError, void>;
+  /**
+   * Return the interaction identfied with the given id.
+   */
+  readonly find: (
+    id: InteractionRequestId
+  ) => TE.TaskEither<DomainError, Option<InteractionRequest>>;
```
Comment (kin0992) [inline src/core/repositories/InteractionRequestRepository.ts:30]:
```suggestion
  readonly upsert: (
    grant: InteractionRequest
  ) => TaskEither<DomainError, InteractionRequest>;
  /**
   * Delete the InteractionRequest identified with the given id.
   */
  readonly remove: (
    id: InteractionRequestId
  ) => TaskEither<DomainError, void>;
  /**
   * Return the interaction identified with the given id.
   */
  readonly find: (
    id: InteractionRequestId
  ) => TaskEither<DomainError, Option<InteractionRequest>>;
```

File: src/core/repositories/SessionRepository.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { Option } from "fp-ts/Option";
+import * as TE from "fp-ts/TaskEither";
```
Comment (kin0992) [inline src/core/repositories/SessionRepository.ts:2]:
```suggestion
import { TaskEither } from "fp-ts/TaskEither";
```

File: src/core/repositories/SessionRepository.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+   * Return the session identfied with the given id.
+   */
+  readonly find: (id: SessionId) => TE.TaskEither<DomainError, Option<Session>>;
+
+  /**
+   * Return the session identfied with the given uid.
+   */
+  readonly findByUid: (
+    id: string
+  ) => TE.TaskEither<DomainError, Option<Session>>;
```
Comment (kin0992) [inline src/core/repositories/SessionRepository.ts:27]:
```suggestion
  readonly upsert: (entity: Session) => TaskEither<DomainError, Session>;
  /**
   * Delete the session identified with the given id.
   */
  readonly remove: (id: SessionId) => TaskEither<DomainError, void>;
  /**
   * Return the session identified with the given id.
   */
  readonly find: (id: SessionId) => TaskEither<DomainError, Option<Session>>;

  /**
   * Return the session identified with the given uid.
   */
  readonly findByUid: (
    id: string
  ) => TaskEither<DomainError, Option<Session>>;
```

File: src/implementations/postgres/grantRepository.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  (logger: Logger) =>
+  <T>(client: Prisma.GrantDelegate<T>) =>
+  (id: GrantId) =>
+    pipe(
+      TE.tryCatch(() => client.findUnique({ where: { id } }), E.toError),
+      TE.map(flow(O.fromNullable, O.map(fromRecord))),
+      TE.mapLeft((e) => ({
+        causedBy: e,
+        kind: DomainErrorTypes.GENERIC_ERROR,
+      })),
```
Comment (kin0992) [inline src/implementations/postgres/grantRepository.ts:70]: We can use a bimap here

File: src/implementations/postgres/interactionRequestRepository.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+          O.map(fromRecord),
+          O.sequence(E.Applicative),
+          E.mapLeft(makeDomainError),
+          TE.fromEither
+        )
+      ),
+      TE.mapLeft((e) => ({
+        causedBy: e,
+        kind: DomainErrorTypes.GENERIC_ERROR,
+      })),
```
Comment (kin0992) [inline src/implementations/postgres/interactionRequestRepository.ts:132]: Same as above, it is possible to use a bimap here

File: src/implementations/postgres/sessionRepository.ts [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+  (logger: Logger) =>
+  <T>(client: Prisma.SessionDelegate<T>) =>
+  (id: SessionId) =>
+    pipe(
+      TE.tryCatch(() => client.findUnique({ where: { id } }), E.toError),
+      TE.map(flow(O.fromNullable, O.map(fromRecord))),
+      TE.mapLeft((e) => ({
+        causedBy: e,
+        kind: DomainErrorTypes.GENERIC_ERROR,
+      })),
```
Comment (kin0992) [inline src/implementations/postgres/sessionRepository.ts:65]: bimap here too

Comment (kin0992) [reply src/implementations/postgres/sessionRepository.ts:65]: and belove as well

**Comment by kin0992** [comment]
You can remove `ioredis` from the dependencies too

File: src/core/repositories/GrantRepository.ts (hunk 2 above)
Comment (datalek) [reply src/core/repositories/GrantRepository.ts]: Done!

File: src/core/repositories/InteractionRequestRepository.ts (hunk 3 above)
Comment (datalek) [reply src/core/repositories/InteractionRequestRepository.ts]: Done!

File: src/core/repositories/InteractionRequestRepository.ts (hunk 4 above)
Comment (datalek) [reply src/core/repositories/InteractionRequestRepository.ts]: Done!

File: src/core/repositories/SessionRepository.ts (hunk 6 above)
Comment (datalek) [reply src/core/repositories/SessionRepository.ts]: Done!

File: src/core/repositories/SessionRepository.ts (hunk 5 above)
Comment (datalek) [reply src/core/repositories/SessionRepository.ts]: Done!

File: src/core/repositories/GrantRepository.ts (hunk 1 above)
Comment (datalek) [reply src/core/repositories/GrantRepository.ts]: Done!

**Review by kin0992 — APPROVED** [review]
🪜

**Pull request merged**